File: Third Year/COS 314/Practicals/Practical 3/NeuralNetwork.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class NeuralNetwork {

    private int inputSize, hiddenSize, outputSize;
    private double learningRate;
    private double[][] W1, W2;
    private double[] b1, b2;
    private double[] z1, a1, z2, output;

    public NeuralNetwork(int inputSize, int hiddenSize, int outputSize, double learningRate) {
        this.inputSize = inputSize;
        this.hiddenSize = hiddenSize;
        this.outputSize = outputSize;
        this.learningRate = learningRate;
        initializeWeights();
    }

    private void initializeWeights() {
        W1 = new double[hiddenSize][inputSize];
        W2 = new double[outputSize][hiddenSize];
        b1 = new double[hiddenSize];
        b2 = new double[outputSize];
        z1 = new double[hiddenSize];
        a1 = new double[hiddenSize];
        z2 = new double[outputSize];
        output = new double[outputSize];

        Random rand = new Random();
        for (int i = 0; i < hiddenSize; ++i) {
            for (int j = 0; j < inputSize; ++j) {
                W1[i][j] = rand.nextDouble() - 0.8;
            }
        }

        for (int i = 0; i < outputSize; ++i) {
            for (int j = 0; j < hiddenSize; ++j) {
                W2[i][j] = rand.nextDouble() - 0.8;
            }
        }
    }

    private double sigmoid(double x) {
        return 1.0 / (1.0 + Math.exp(-x));
    }

    private double sigmoid_derivative(double x) {
        double s = sigmoid(x);
        return s * (1 - s);
    }

    public double[] forwardPass(double[] input) {
        for (int i = 0; i < hiddenSize; ++i) {
            z1[i] = b1[i];
            for (int j = 0; j < inputSize; ++j) {
                z1[i] += W1[i][j] * input[j];
            }
            a1[i] = sigmoid(z1[i]);
        }

        for (int i = 0; i < outputSize; ++i) {
            z2[i] = b2[i];
            for (int j = 0; j < hiddenSize; ++j) {
                z2[i] += W2[i][j] * a1[j];
            }
            output[i] = sigmoid(z2[i]);
        }

        return output;
    }

    public void backwardPass(double[] input, double[] target) {
        double[] delta2 = new double[outputSize];
        for (int i = 0; i < outputSize; ++i) {
            delta2[i] = (output[i] - target[i]) * sigmoid_derivative(z2[i]);
        }

        double[] delta1 = new double[hiddenSize];
        for (int i = 0; i < hiddenSize; ++i) {
            double error = 0.0;
            for (int j = 0; j < outputSize; ++j) {
                error += delta2[j] * W2[j][i];
            }
            delta1[i] = error * sigmoid_derivative(z1[i]);
        }

        for (int i = 0; i < outputSize; ++i) {
            for (int j = 0; j < hiddenSize; ++j) {
                W2[i][j] -= learningRate * delta2[i] * a1[j];
            }
            b2[i] -= learningRate * delta2[i];
        }

        for (int i = 0; i < hiddenSize; ++i) {
            for (int j = 0; j < inputSize; ++j) {
                W1[i][j] -= learningRate * delta1[i] * input[j];
            }
            b1[i] -= learningRate * delta1[i];
        }
    }

    public void train(List<double[]> inputs, List<double[]> targets, int epochs) {
        for (int epoch = 0; epoch < epochs; ++epoch) {
            for (int i = 0; i < inputs.size(); ++i) {
                forwardPass(inputs.get(i));
                backwardPass(inputs.get(i), targets.get(i));
            }
        }
    }

    public static void main(String[] args) throws IOException {
        // Define input data and target values based on the given data
        List<double[]> inputs = new ArrayList<>();
        List<double[]> targets = new ArrayList<>();

        // Load data from text file
        BufferedReader inputFile = new BufferedReader(new FileReader("traindata.txt"));
        String line;
        while ((line = inputFile.readLine()) != null) {
            String[] tokens = line.split(",");
            double[] input = new double[tokens.length - 1];
            for (int i = 0; i < tokens.length - 1; i++) {
                input[i] = Double.parseDouble(tokens[i]);
            }
            inputs.add(input);

            double[] target = {Double.parseDouble(tokens[tokens.length - 1])};
            targets.add(target);
        }
        inputFile.close();

        // Create neural network: 8 inputs, 3 hidden neurons, 1 output, learning rate 0.1
        NeuralNetwork nn = new NeuralNetwork(2, 10, 1, 0.8);

        // Train neural network
        nn.train(inputs, targets, 300);

        // Test neural network
        for (int i = 0; i < inputs.size(); i++) {
            double[] input = inputs.get(i);
            double[] output = nn.forwardPass(input);
            System.out.print("Input: ");
            for (double val : input) {
                System.out.print(val + " ");
            }
            System.out.println(" => Output: " + output[0] + " Expected: " + targets.get(i)[0]);
        }
    }
}

File: Third Year/COS 314/Practicals/Practical 3/NeuralNetwork.cpp
#include "NeuralNetwork.h"
#include "../../../../../../../../../../usr/include/c++/11/bits/algorithmfwd.h"

using namespace std;

NeuralNetwork::NeuralNetwork(int inputSize, int hiddenSize, int outputSize, double learningRate)
    : inputSize(inputSize), hiddenSize(hiddenSize), outputSize(outputSize), learningRate(learningRate)
{
    initializeWeights();
}

void NeuralNetwork::Run()
{
    // srand(1);
    srand(123456);

    pair<vector<vector<double>>, vector<vector<double>>> data = readData("traindata.txt");
    vector<vector<double>> inputs = data.first;
    vector<vector<double>> targets = data.second;

    // Shuffle the inputs
    random_shuffle(inputs.begin(), inputs.end());

    train(inputs, targets, 10);

    pair<vector<vector<double>>, vector<vector<double>>> testdata = readData("testdata.txt");
    inputs = testdata.first;
    targets = testdata.second;
    cout << "Testing the neural network" << endl;

    int TP = 0, TN = 0, FP = 0, FN = 0;

    for (size_t i = 0; i < inputs.size(); ++i)
    {
        vector<double> output = forwardPass(inputs[i]);
        double predictedOutput = output[0];

        // Threshold the predicted output
        if (predictedOutput > 0.5)
        {
            predictedOutput = 1;
        }
        else
        {
            predictedOutput = 0;
        }

        double actualOutput = targets[i][0];

        if (predictedOutput == actualOutput)
        {
            if (predictedOutput == 1)
            {
                ++TP;
            }
            else
            {
                ++TN;
            }
        }
        else
        {
            if (predictedOutput == 1)
            {
                ++FP;
            }
            else
            {
                ++FN;
            }
        }

        
    }

    double accuracy = static_cast<double>(TP + TN) / (TP + TN + FP + FN);
    double specificity = static_cast<double>(TN) / (TN + FP);
    double sensitivity = static_cast<double>(TP) / (TP + FN);
    double precision = static_cast<double>(TP) / (TP + FP);
    double fMeasure = 2 * (precision * sensitivity) / (precision + sensitivity);

    cout << "Accuracy: " << accuracy * 100 << "%" << endl;
    cout << "Specificity: " << specificity << endl;
    cout << "Sensitivity: " << sensitivity << endl;
    cout << "F-measure: " << fMeasure << endl;
}

void NeuralNetwork::initializeWeights()
{
    // cout << "Initializing weights" << endl;
    // Initialize weights
    Weights.push_back(vector<double>(inputSize * hiddenSize));
    Weights.push_back(vector<double>(hiddenSize * outputSize));

    for (size_t i = 0; i < Weights.size(); ++i)
    {
        for (size_t j = 0; j < Weights[i].size(); ++j)
        {
            Weights[i][j] = randomWeight();
        }
    }
}

vector<double> NeuralNetwork::forwardPass(const vector<double> &input)
{
    vector<double> hiddenLayer(hiddenSize);
    vector<double> outputLayer(outputSize);

    // Calculate the hidden layer
    for (int i = 0; i < hiddenSize; ++i)
    {
        double sum = bias * Weights[0][i];
        for (int j = 0; j < inputSize; ++j)
        {
            sum += input[j] * Weights[0][i * inputSize + j];
        }
        hiddenLayer[i] = sigmoid(sum);
    }

    // Calculate the output layer
    for (int i = 0; i < outputSize; ++i)
    {
        double sum = bias * Weights[1][i];
        for (int j = 0; j < hiddenSize; ++j)
        {
            sum += hiddenLayer[j] * Weights[1][i * hiddenSize + j];
        }
        outputLayer[i] = sigmoid(sum);
    }

    return outputLayer;
}

void NeuralNetwork::backwardPass(const vector<double> &input, const vector<double> &target)
{
    vector<double> hiddenLayer(hiddenSize);
    vector<double> outputLayer(outputSize);

    // Calculate the hidden layer
    for (int i = 0; i < hiddenSize; ++i)
    {
        double sum = 0.0;
        for (int j = 0; j < inputSize; ++j)
        {
            sum += input[j] * Weights[0][i * inputSize + j];
        }
        hiddenLayer[i] = sigmoid(sum);
    }

    // Calculate the output layer
    for (int i = 0; i < outputSize; ++i)
    {
        double sum = 0.0;
        for (int j = 0; j < hiddenSize; ++j)
        {
            sum += hiddenLayer[j] * Weights[1][i * hiddenSize + j];
        }
        outputLayer[i] = sigmoid(sum);
    }

    // Calculate the output layer error
    vector<double> outputError(outputSize);
    for (int i = 0; i < outputSize; ++i)
    {
        outputError[i] = (target[i] - outputLayer[i]) * sigmoid_derivative(outputLayer[i]);
    }

    // Calculate the hidden layer error
    vector<double> hiddenError(hiddenSize);
    for (int i = 0; i < hiddenSize; ++i)
    {
        double sum = 0.0;
        for (int j = 0; j < outputSize; ++j)
        {
            sum += Weights[1][j * hiddenSize + i] * outputError[j];
        }
        hiddenError[i] = sum * sigmoid_derivative(hiddenLayer[i]);
    }

    // Update the weights of the output layer
    for (int i = 0; i < outputSize; ++i)
    {
        for (int j = 0; j < hiddenSize; ++j)
        {
            Weights[1][i * hiddenSize + j] += learningRate * outputError[i] * hiddenLayer[j];
        }
    }

    for (int i = 0; i < hiddenSize; ++i)
    {
        for (int j = 0; j < inputSize; ++j)
        {
            Weights[0][i * inputSize + j] += learningRate * hiddenError[i] * input[j];
        }
    }

    // Update the bias
    bias += learningRate * hiddenError[0];

    // output the output layer error
    // cout << "Output layer error: " << outputError[0] << endl;
}

void NeuralNetwork::train(const vector<vector<double>> &inputs, const vector<vector<double>> &targets, int epochs)
{
    for (int epoch = 0; epoch < epochs; ++epoch)
    {
        // cout << "Epoch " << epoch << endl;
        // cout << "Inputs size: " << inputs.size() << endl;
        for (size_t i = 0; i < inputs.size(); ++i)
        {
            // cout << "Forward pass" << endl;
            forwardPass(inputs[i]);
            // cout << "Backward pass" << endl;
            backwardPass(inputs[i], targets[i]);
        }
    }
}

vector<string> NeuralNetwork::split(const string &str, char delimiter)
{
    vector<string> tokens;
    string token;
    stringstream ss(str); // Use a stringstream to parse the string

    while (getline(ss, token, delimiter))
    {
        tokens.push_back(token); // Store each token in the vector
    }

    return tokens;
}

pair<vector<vector<double>>, vector<vector<double>>> NeuralNetwork::readData(string filename)
{
    ifstream file(filename);
    string line;
    vector<vector<double>> inputs;
    vector<vector<double>> targets;

    while (getline(file, line))
    {
        vector<string> tokens = split(line, ',');
        vector<double> input;
        vector<double> target;

        for (size_t i = 0; i < tokens.size(); ++i)
        {
            input.push_back(stod(tokens[i]));
        }
        targets.push_back({input.back()});
        input.pop_back();
        inputs.push_back(input);
    }

    return make_pair(inputs, targets);
}

